### PR TITLE
Add loading and building for OSX driver

### PR DESCRIPTION
**Pull request by tweksteen**

### CODE REVIEW

File: chipsec/helper/osx/helper.py [hunk 1]
```diff
@@ -48,6 +51,23 @@ def __init__(self):
...
+                                   "helper", "osx", self.DRIVER_NAME)
+        # Make sure the driver image and its subdirectories are owned by root.
+        s = os.stat(driver_path)
+        if s.st_uid != 0 or s.st_gid != 0:
+            os.chown(driver_path, 0, 0)
+            for root, dirs, files in os.walk(driver_path):
+                for f in dirs + files:
+                    os.chown(os.path.join(root, f), 0, 0)
+        subprocess.check_call(["kextload", driver_path])
+        if os.path.exists(self.DEVICE_NAME):
```
Comment (iplosion) [inline chipsec/helper/osx/helper.py:65]: I think you should look at the return code of the subprocess call to make a decision if the module is loaded or not.

File: setup.py [hunk 2]
```diff
@@ -64,26 +65,45 @@ def finalize_options(self):
...
+        # Run the makefile there.
+        subprocess.check_output(["make", "-C", build_driver])
+        # And copy the resulting .ko to the right place.
+        # That is to the source directory if we are in "develop" mode,
+        # otherwise to the helper subdirectory in the build directory.
+        root_dst = "" if self.inplace else self.real_build_lib
+        dst = os.path.join(root_dst, "chipsec", "helper", "linux")
+        self.copy_file(ko_ext, dst)
+        # Finally, we clean up the build directory.
+        shutil.rmtree(os.path.join(self.real_build_lib, "drivers"))
```
Comment (iplosion) [inline setup.py]: Why are you using shutil.rmtree in this case? You are already working with dist_util which offers a remove_tree function?

File: setup.py [hunk 3]
```diff
@@ -64,26 +65,45 @@ def finalize_options(self):
...
+
+    def _build_darwin_driver(self):
+        log.info("building the OSX driver")
+        build_driver = os.path.join(self.real_build_lib, "drivers", "osx")
+        xcodeproject = os.path.join(build_driver, "chipsec.xcodeproj")
+        self.copy_tree(os.path.join("drivers", "osx"), build_driver)
+        subprocess.check_output(["xcodebuild", "-project", xcodeproject,
```
Comment (iplosion) [inline setup.py:92]: Same here in regards to error handling.

File: setup.py [hunk 4]
```diff
@@ -64,26 +65,45 @@ def finalize_options(self):
...
+        # And copy the resulting .ko to the right place.
+        # That is to the source directory if we are in "develop" mode,
+        # otherwise to the helper subdirectory in the build directory.
+        root_dst = "" if self.inplace else self.real_build_lib
+        dst = os.path.join(root_dst, "chipsec", "helper", "linux")
+        self.copy_file(ko_ext, dst)
+        # Finally, we clean up the build directory.
+        shutil.rmtree(os.path.join(self.real_build_lib, "drivers"))
+
+    def _build_darwin_driver(self):
```
Comment (iplosion) [inline setup.py:85]: Add some comments for the steps as you done in the Linux function.

File: setup.py [hunk 5]
```diff
@@ -64,26 +65,45 @@ def finalize_options(self):
...
-            dst = os.path.join(root_dst, "chipsec", "helper", "linux")
-            self.copy_file(ko_ext, dst)
-            # Finally, we clean up the build directory.
-            shutil.rmtree(os.path.join(build_lib, "drivers"))
+        if not self.skip_driver:
+            self.real_build_lib = os.path.realpath(self.build_lib)
+            if platform.system().lower() == "linux":
+                self._build_linux_driver()
+            elif platform.system().lower() == "darwin":
+                self._build_darwin_driver()
```
Comment (iplosion) [inline setup.py:112]: Is this okay if there is no else path, let's say it is Windows?

**Review by iplosion — COMMENTED** [review]
I am not familiar with setup tools, so some comments are related to my lack of knowledge.

File: chipsec/helper/osx/helper.py (hunk 1 above)
Comment (tweksteen) [reply chipsec/helper/osx/helper.py:65]: `check_call` is already looking at the return code. This is an extra check.

File: setup.py (hunk 2 above)
Comment (tweksteen) [reply setup.py]: Done.

File: setup.py (hunk 4 above)
Comment (tweksteen) [reply setup.py:85]: Done.

File: setup.py (hunk 3 above)
Comment (tweksteen) [reply setup.py:92]: Same here. `check_output` will verify the value of the return code and raise CalledProcessError if it failed.

File: setup.py (hunk 5 above)
Comment (tweksteen) [reply setup.py:112]: Yes. For now, the driver is not built on windows but just packaged.

**Comment by iplosion** [comment]
LGTM now.

File: chipsec/helper/osx/helper.py [hunk 6]
```diff
@@ -48,6 +51,23 @@ def __init__(self):
...
+            if logger().VERBOSE:
+                logger().log("Module %s loaded successfully" % self.DRIVER_NAME)
+        else:
+            logger().error("Fail to load the module %s" % self.DRIVER_NAME)
```
Comment (af3556) [inline chipsec/helper/osx/helper.py]: s/Fail/Failed/

Comment (af3556) [reply chipsec/helper/osx/helper.py]: lgtm

Comment (tweksteen) [reply chipsec/helper/osx/helper.py]: Done. Thanks.

**Pull request merged**